### DISCUSSION TITLE
How does llama.cpp manage the memory?

### DISCUSSION THREAD

**Discussion by VincentXWD** [ggml-org/llama.cpp#6324] [General]
Hello, I'm wondering how does llama.cpp manage the memory:
1. Does llama.cpp allocate space for tensors including static parameter tensors and temporary tensors at once? I only accumulated the allocations of parameter tensors(e.g. **blk.0.attn_q.weight**) but no more temporary tensors (e.g. **inp_embd**). Could you please explain where the allocating process is?
2. Will llama.cpp free all tensors at the end of the model inference process including both parameter tensors and temporary tensors? I'm also wondering where llama.cpp frees these tensors.

I hope I described my confusions properly and thanks for your attention.

**Comment by slaren** [comment]
Hi, the inputs and temporary tensors used in the computation graph are allocated by `ggml-alloc` through `ggml_backend_sched`. This is a custom allocator that seeks to reduce memory usage by delaying allocation of the tensors until they are used, and freeing them when they are not needed any longer.

**Reply by VincentXWD** [reply]
Thank you for your quick reply!

**Comment by VincentXWD** [comment]
Hi slaren,
I investigated the `ggml_backend_sched` and it seems like to allocate the memory buffer at once (in cuda, kompute and etc.). I still cannot understand the logic. Could you please give me a brief introduction to the memory allocating mechanism of any one of these backends? This will help me and other users understand llama.cpp more in memory management.

Thank you.

**Reply by slaren** [reply]
`ggml_backend_sched` adds some complexity to the allocator to support multiple devices, but it does not perform any allocation by itself. You should be looking at `ggml_gallocr` (graph allocator) in `ggml-alloc.c`.

After the computation graph is built, llama.cpp creates a list of nodes in topological order that will be executed in sequence. Functionally, what the allocator does is roughly equivalent to this:

```python
# initialize node.refs to the number of times it appears as a source in the graph
for node in nodes:
  allocate(node)
  execute(node)
  for src in node.sources:
    src.refs -= 1
    if src.refs == 0: free(src)
```

However, instead of doing this allocation in real time as the graph is executed, `ggml-alloc` does it in a preprocessing step, where the offsets within the compute buffer where each tensor will be allocated is calculated.

**Reply by VincentXWD** [reply]
yes, i found it. this is a dynamic allocating strategy. thank you so much!

**Reply by fiskrt** [reply]
Hi @slaren, can you expand more on the memory allocation strat?

If I run a model with `-n 32 --no-warmup --log-verbose -c 1024`
I'm seeing the following before the graph is executed:
```
// Allocate the KV cache. This makes sense
llama_kv_cache_unified: layer   0,1,2,3...: dev = Metal
// reserve pp graph first so that buffers are only allocated once
graph_reserve: reserving a graph for ubatch with n_tokens =  512, n_seqs =  1, n_outputs =  512
ggml_dyn_tallocr_alloc: allocating inp_embd (6291456 bytes) - 
... alloc rest of tensors needed for forward pass for 512 token inp
// reserve with tg graph to get the number of splits and nodes
graph_reserve: reserving a graph for ubatch with n_tokens =    1, n_seqs =  1, n_outputs =    1
ggml_dyn_tallocr_alloc: allocating inp_embd (12288 bytes) - 
... alloc rest of tensors needed for forward pass for 1 token inp
// reserve again with pp graph to avoid ggml-alloc reallocations during inference
graph_reserve: reserving a graph for ubatch with n_tokens =  512, n_seqs =  1, n_outputs =  512
ggml_dyn_tallocr_alloc: allocating inp_embd (6291456 bytes) - 
... alloc rest of tensors needed for forward pass for 512 token inp
Execute graph ....
```
1. Can you explain the comments and why we init PP twice?
2. Also is the strategy really dynamic? It seems like all the tensors are allocated before execution, for example reusing some buffers with inplace computations when possible. How does this align with the pseudo-code you gave above?
3. How does this relate to the mmaping and moving from other devices such as Metal/CUDA?

**Reply by slaren** [reply]
- The reason we init twice is explained in the code:
```C++
// reserve pp (prompt processing) graph first so that buffers are only allocated once
// [...]
// reserve with tg (token generation) graph to get the number of splits and nodes
// [...]
// reserve again with pp graph to avoid ggml-alloc reallocations during inference
// [...]
```

- The tensors are pre-allocated using this strategy. The allocation strategy is the same as the pseudocode, with the exception that the nodes are not executed at that point.

- mmaping doesn't change this. Using multiple devices in the same graph complicates the process a bit, but fundamentally the algorithm is the same.

**Comment by Zijie-Tian** [comment]
> After the computation graph is built, llama.cpp creates a list of nodes in topological order that will **be executed in sequence.** Functionally, what the allocator does is roughly equivalent to this:

@slaren Hi slaren, I wonder if the "execute" of this graph is an asynchronous process. If one part of the node is in the CPU and the other part is in the GPU, is there a possibility of parallelism? Is there any issue or discussion about this?

**Reply by slaren** [reply]
It is possible if there are no dependencies between the parts, that's how pipeline parallelism is implemented. However, the CPU backend does not support asynchronous computation, so it is always run synchronously.

**Reply by Zijie-Tian** [reply]
OK, I see. So for the current version of llama.cpp, there is no way for me to execute the independent tensor of one layer on both CPU and GPU, just like tensor parallelism? 
As you know, for Edge model inference, executing sequentially within the SoC would waste compute resources. Does llama.cpp have some plans for this issue?

**Reply by slaren** [reply]
It doesn't really waste any resources because pipeline parallelism can only be used with large batches, and large batches are offloaded to the GPU instead since that gives much better performance. During generation (batch size=1), there is no opportunity for parallel computation since the GPU part depends on the output of the CPU part.

**Reply by Zijie-Tian** [reply]
![image](https://github.com/user-attachments/assets/bdd07e97-b4f8-4101-b3d1-68840891c22c)
As shown in this  #9086 , what I mean is splitting the operation between the CPU and GPU. Specifically, dividing the MUL_MAT operation so that one part executes on the CPU and the other on the GPU, and add a concat node. Is there potential for performance improvement?

**Reply by slaren** [reply]
This kind of tensor parallelism between different backends is not supported. It may be possible to improve generation performance in this way, but there could be a significant synchronization overhead. @JohannesGaessler mentioned the possibility of reducing the synchronization overhead in the attention by computing each attention head in a different device.

**Reply by Zijie-Tian** [reply]
OK, Thanks for the response.

**Reply by JohannesGaessler** [reply]
>what I mean is splitting the operation between the CPU and GPU. Specifically, dividing the MUL_MAT operation so that one part executes on the CPU and the other on the GPU, and add a concat node. Is there potential for performance improvement?

When I initially did the implementation of what is currently `--split-mode row` I tested parallelizing CPU and GPU but the overhead from CPU <-> GPU copies and synchronization seemed to have been much larger than any potential benefit. At the time there was also no support for CUDA graphs whatsoever though which may help with at least the synchronization overhead. My overall intuition is however that it would still not be viable.

**Reply by Zijie-Tian** [reply]
I think, for SoCs (like Jetson or Mac devices), when using cudaManagedMemory, the copy overhead between the CPU and GPU seems negligible. Also, we only need to use ManagedMemory for the activation variables, and it appears the current llama.cpp implementation hasn’t optimized in this area yet.

**Reply by Zijie-Tian** [reply]
> At the time there was also no support for CUDA graphs whatsoever though which may help with at least the synchronization overhead. My overall intuition is however that it would still not be viable.

Yes, you're right; cudaGraph indeed cannot be used.

**Reply by JohannesGaessler** [reply]
No, what I meant to say is that if I remember correctly CUDA graphs have functionality where you can embed CPU code which I would expect to have less overhead than doing memcpys and calling e.g. `cudaStreamSynchronize`.

**Reply by Zijie-Tian** [reply]
OK, Thanks for the response; I'll give it a try. If there are good results, we can discuss this issue elsewhere.

**Reply by DavidZyy** [reply]
@Zijie-Tian Is there any follow-up to this matter? I want to know the result.
